File: nim/Day6.nim
import sets, strformat

proc solution(window: int): int =
    let file = open("input/Day6.txt");
    defer: file.close()

    for i in file.lines():
        for j in 0..<len(i) - window - 1:
            if len(i[j..<j + window].toHashSet()) == window:
                return j + window
let 
    ttl1 = solution(4)
    ttl2 = solution(14)

echo &"Solution 1: {ttl1}\nSolution 2: {ttl2}"

File: rust/day6.rs
use std::collections::HashSet;

fn main() {
    let input = std::fs::read_to_string("../inputs/day6.txt").expect("Incorrect File Path");
    let pt1 = solution(&input, 4);
    let pt2 = solution(&input, 14);
    println!("Part 1: {}\nPart 2: {}", pt1, pt2)
}

fn solution(input: &str, window: usize) -> usize {
    input
        .split("\r\n\r\n")
        .flat_map(|line| line.lines())
        .map(|s| {
            (0..=s.len().saturating_sub(window))
                .find(|&i| s[i..i + window].chars().collect::<HashSet<char>>().len() == window)
                .map_or(0, |i| i + window)
        })
        .sum::<usize>()
}
